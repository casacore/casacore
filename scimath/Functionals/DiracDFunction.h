//# DiracDFunction.h: A one dimensional Dirac delta function
//# Copyright (C) 2002,2005
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#ifndef SCIMATH_DIRACDFUNCTION_H
#define SCIMATH_DIRACDFUNCTION_H

//# Includes
#include <casa/aips.h>
#include <scimath/Functionals/DiracDParam.h>
#include <scimath/Functionals/Function.h>

namespace casa { //# NAMESPACE CASA - BEGIN

//# Forward declarations

// <summary> A one dimensional Dirac delta function
// </summary>

// <use visibility=export>

// <reviewed reviewer="" date="" tests="tFunctionHolder" 
// demos="">
// </reviewed>

// <prerequisite>
//   <li> <linkto class="DiracDParam">DiracDParam</linkto>
//   <li> <linkto class="Function">Function</linkto>
// </prerequisite>

// <etymology> 
// A 1-dimensional Dirac delta.
// </etymology>

// <synopsis> 
// A <src>DiracD</src> is described by a height, a center and a width
// (halfwidth). The value is:
// <srcblock>
//      height          (|x-center| == 0.0)
//      0               (|x-center| != 0.0)
// </srcblock>
// The parameters are enumerated by HEIGHT, CENTER and WIDTH. They have
// default values of (1, 0).
// </synopsis> 
//
// <example>
// <srcblock>
//    DiracDFunction<Double> sf(5.0, 25.0);
//    sf(25);            // = 5.0
// </srcblock>
// </example>

// <templating arg=T>
//  <li> T should have standard numerical operators
// </templating>

// <thrown>
//    <li> AipsError if incorrect parameter number specified.
// </thrown>
//

template<class T> class DiracDFunction : public DiracDParam<T>
{
public:
  //# Constructors
  // Constructs the DiracDFunction, Defaults:
  // height=1, center=0.
  // <note role=warning> Could not use default arguments
  // that worked both with gcc and IRIX </note>
  // <group>
  DiracDFunction() : DiracDParam<T>() {};
  explicit DiracDFunction(const T &height) :
    DiracDParam<T>(height) {};
  DiracDFunction(const T &height, const T &center) :
    DiracDParam<T>(height, center) {};
  // </group>

  // Copy constructor (deep copy)
  // <group>
  DiracDFunction(const DiracDFunction<T> &other) : DiracDParam<T>(other) {};
  template <class W>
  DiracDFunction(const DiracDFunction<W> &other) : DiracDParam<T>(other) {};
  // </group>

  // Copy assignment (deep copy)
  DiracDFunction<T> &operator=(const DiracDFunction<T> &other) {
    DiracDParam<T>::operator=(other); return *this; };
    
  // Destructor
  virtual ~DiracDFunction() {};

  //# Operators    
  // Evaluate the DiracD at <src>x</src>.
  // If a vector is used as the argument only its first element is used.
  // <group>
  virtual T eval(typename Function<T>::FunctionArg x) const;
  // </group>
    
  //# Member functions
  // Return a copy of this object from the heap. The caller is responsible 
  // for deleting this pointer. 
  // <group>
  virtual Function<T> *clone() const { return new DiracDFunction<T>(*this); };
  virtual Function<typename FunctionTraits<T>::DiffType> *cloneAD() const {
    return new DiracDFunction<typename FunctionTraits<T>::DiffType>(*this); };
  virtual Function<typename FunctionTraits<T>::BaseType> *cloneNonAD() const {
    return new DiracDFunction<typename FunctionTraits<T>::BaseType>(*this); };
  // </group>

  //# Make members of parent classes known.
protected:
  using DiracDParam<T>::param_p;
public:
  using DiracDParam<T>::nparameters;
  using DiracDParam<T>::HEIGHT;
  using DiracDParam<T>::CENTER;
};


} //# NAMESPACE CASA - END

#ifndef CASACORE_NO_AUTO_TEMPLATES
#include <scimath/Functionals/DiracDFunction.tcc>
#endif //# CASACORE_NO_AUTO_TEMPLATES
#endif
