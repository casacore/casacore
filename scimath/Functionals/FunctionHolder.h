//# FunctionHolder.h: A holder for Functions to enable record conversions
//# Copyright (C) 2002,2003,2004
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#ifndef SCIMATH_FUNCTIONHOLDER_H
#define SCIMATH_FUNCTIONHOLDER_H

//# Includes
#include <casa/aips.h>
#include <casa/Utilities/PtrHolder.h>
#include <casa/Utilities/RecordTransformable.h>
#include <scimath/Functionals/Function.h>
#include <casa/Arrays/Vector.h>
#include <casa/BasicSL/String.h>

namespace casa { //# NAMESPACE CASA - BEGIN

//# Forward Declarations

// <summary> A holder for Functions to enable record conversions </summary>

// <use visibility=export>

// <reviewed reviewer="" date="yyyy/mm/dd" tests="tFunctionHolder" demos="">
// </reviewed>

// <prerequisite>
//   <li> <linkto class=RecordInterface>RecordInterface</linkto> class
//   <li> <linkto class=Function>Function</linkto> class
// </prerequisite>
//
// <etymology>
// A Holder of general Measures
// </etymology>
//
// <synopsis>
// This class can be used to handle heterogeneous collections of Functions,
// e.g. as a <src>Vector<FunctionHolder></src>. With the aid of the
// toRecord() and fromRecord() functions it can be used
// to convert a Function object into or from a record.
// A FunctionHolder is created from a Function, or can be empty.
//
// </synopsis>
//
// <example>
// <srcblock>
//	TableRecord rec;
//	MDirection dir(MVDirection(Quantity(12.5, 'deg'), Quantity(-2, 'deg')),
//		       MDirection::J2000);
//	String error;		// error message
//	if (!FunctionHolder(dir).toRecord(error, rec)) {
//		cout << error << endl;
//	};
//	Record grec;		// a Record
//	if (!FunctionHolder(dir).toRecord(error, grec)) {  // make record
//		cout << error << endl;
//	};
// // Note that for GlishRecords use can be made of the
// // GlishRecord::to/fromrecord() methods.
// </srcblock>
// </example>
//
// <motivation>
// To make general conversions between Functions and records, without knowing
// the actual Function being converted.
// </motivation>

template <class T> class FunctionHolder : public RecordTransformable {
 public:
  //# Enumerations
  // Types of functions
  enum Types {
    GAUSSIAN1D,
    GAUSSIAN2D,
    GAUSSIAN3D,
    GAUSSIANND,
    HYPERPLANE,
    POLYNOMIAL,
    EVENPOLYNOMIAL,
    ODDPOLYNOMIAL,
    SINUSOID1D,
    CHEBYSHEV,
    BUTTERWORTH,
    COMBINE,
    COMPOUND,
    COMPILED,
    N_Types
  };
  //# Structures
  // Structure to hold functional status
  struct FuncStat {
    // Name
    String nam;
    // type
    Types tp;
    // Order (True if needed)
    Bool order;
  };

  //# Constructors
  // Creates an empty holder
  FunctionHolder();
  // Create from a Function (copy made)
  FunctionHolder(const Function<T> &in);
  // Copy a holder (copy semantics)
  FunctionHolder(const FunctionHolder<T> &other);
  //# Destructor
  ~FunctionHolder();

  //# Operators
  // Assignment (copy semantics)
  FunctionHolder &operator=(const FunctionHolder<T> &other);

  //# Member Functions
  // Check the the FunctionHolder holds the specified type. Return
  // True if if does and False otherwise.
  // <group>
  Bool isEmpty() const;
  // </group>
  // Get the known names
  const Vector<String> &names() const;
  // Get a specific Function from the holder (with lifetime as long 
  // as holder exists).
  // <thrown>
  // <li> AipsError if holder empty
  // <li> AipsError if holder contains wrong Function
  // </thrown>
  // <group>
  const Function<T> &asFunction() const;
  // </group>
  // Add a function
  Bool addFunction(const Function<T> &fnc); 
  // Get the type of currently filled holder
  Types type() const;
  // Create a Function from a record. An error message is generated, and False
  // returned if an invalid record is given. A valid record will return True.
  // A valid record contains at least the following fields (any additional fields are
  // ignored):
  // <ul>
  // <li> tp = TpString: type of Function (gaussian1d, etc; case
  //	 insensitive) -- OR an enumeration code
  // <li> order = TpInt: the order needed to create a Function (-1 if not
  // 	necessary or default)
  // <li> ndim, npar, params are optional
  // <li> nfunc, funcs are required for COMBI or COMPOUND
  // </ul>
  // A Function can be created from a string. In that case the string
  // will only indicate the type of function (like polynomial), and will
  // create a default polynomial of that given type. 
  // Error messages are postfixed to error.
  // <group>
  virtual Bool fromRecord(String &error, const RecordInterface &in);
  virtual Bool fromString(String &error, const String &in);
  template <class U>
    Bool getRecord(String &error, Function<U> *&fn,
		   const RecordInterface &in);
  // </group>
  // Create a record from a Function. The return will be False and an error
  // message generated only if the FunctionHolder does not contain a Function.
  // Error messages are postfixed to error.
  virtual Bool toRecord(String &error, RecordInterface &out) const;
  // Get identification of record
  virtual const String &ident() const;

private:
  //# Data Members
  // Pointer to a Function
  PtrHolder<Function<T> > hold_p;
  // Aids (only filled after a succesful to/fromRecord
  // <group>
  mutable Types nf_p;
  mutable Int order_p;
  mutable String text_p;
  mutable PtrHolder<RecordInterface> mode_p;
  // </group>
  // List of known names
  mutable Vector<String> nam_p;
  // Filled list?
  mutable Bool isFilled;

  //# Member functions
  // Initialise and check the name list
  void init() const;
  // Aid for to/from Record, String
  // <group>
  Bool putType(String &error, RecordInterface &out) const;
  template <class U>
    Bool getType(String &error, Function<U> *&fn, const RecordInterface &in);  
  template <class U>
    Bool getType(String &error, Function<U> *&fn);
  void setParameters(Function<T> *&fn, const Vector<T> &params);
  void setParameters(Function<AutoDiff<T> > *&fn, const Vector<T> &params);
  // </group>
};


} //# NAMESPACE CASA - END

#ifndef CASACORE_NO_AUTO_TEMPLATES
#include <scimath/Functionals/FunctionHolder.tcc>
#endif //# CASACORE_NO_AUTO_TEMPLATES
#endif
