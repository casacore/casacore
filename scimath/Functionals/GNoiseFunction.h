//# GNoiseFunction.h: A one dimensional normal distribution 
//# Copyright (C) 2002,2005
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#ifndef SCIMATH_GNOISEFUNCTION_H
#define SCIMATH_GNOISEFUNCTION_H

//# Includes
#include <casa/aips.h>
#include <scimath/Functionals/GNoiseParam.h>
#include <scimath/Functionals/Function.h>

namespace casa { //# NAMESPACE CASA - BEGIN

//# Forward declarations

// <summary> A one dimensional normal distribution 
// </summary>

// <use visibility=export>

// <reviewed reviewer="" date="" tests="tFunctionHolder" 
// demos="">
// </reviewed>

// <prerequisite>
//   <li> <linkto class="GNoiseParam">GNoiseParam</linkto>
//   <li> <linkto class="Function">Function</linkto>
// </prerequisite>

// <etymology> 
// Gaussian Noise generator.
// </etymology>

// <synopsis> 
// A <src>GNoise</src> is described by a mean and a variance (Note these are
// not parameters in the <src>Function</src> sense, but more like the
// order of a polynomial. The defaults are 0 and 1.
// </synopsis> 
//
// <example>
// <srcblock>
//    GNoiseFunction<Double> sf;
//    sf();            // = 0.12
// </srcblock>
// </example>

// <templating arg=T>
//  <li> T should have standard numerical operators
// </templating>

// <thrown>
//    <li> AipsError if incorrect parameter number specified.
// </thrown>
//

template<class T> class GNoiseFunction : public GNoiseParam<T>
{
public:
  //# Constructors
  // Constructs the GNoise, Defaults:
  // mean=0, var=1.0
  // <group>
  GNoiseFunction() : GNoiseParam<T>() {};
  GNoiseFunction(const Double &mean, const Double &var) :
    GNoiseParam<T>(mean, var) {};
  // </group>

  // Copy constructor (deep copy)
  // <group>
  GNoiseFunction(const GNoiseFunction<T> &other) : GNoiseParam<T>(other) {};
  template <class W>
    GNoiseFunction(const GNoiseFunction<W> &other) : GNoiseParam<T>(other) {};
  // </group>

  // Copy assignment (deep copy)
  GNoiseFunction<T> &operator=(const GNoiseFunction<T> &other) {
    GNoiseParam<T>::operator=(other); return *this; };
    
  // Destructor
  virtual ~GNoiseFunction() {};

  //# Operators    
  // Evaluate the GNoise at <src>x</src>.
  // If a vector is used as the argument only its first element is used.
  // <group>
  virtual T eval(typename Function<T>::FunctionArg x) const;
  // </group>
    
  //# Member functions
  // Return a copy of this object from the heap. The caller is responsible 
  // for deleting this pointer. 
  // <group>
  virtual Function<T> *clone() const { return new GNoiseFunction<T>(*this); };
  // </group>

  //# Make members of parent classes known.
protected:
  using GNoiseParam<T>::param_p;
public:
  using GNoiseParam<T>::noise_p;
};


} //# NAMESPACE CASA - END

#ifndef CASACORE_NO_AUTO_TEMPLATES
#include <scimath/Functionals/GNoiseFunction.tcc>
#endif //# CASACORE_NO_AUTO_TEMPLATES
#endif
