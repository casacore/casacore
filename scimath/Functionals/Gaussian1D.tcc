//# Gaussian1D.cc: A one-dimensional Gaussian class
//# Copyright (C) 1994,1995,1996,1999,2001,2002
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#ifndef SCIMATH_GAUSSIAN1D_TCC
#define SCIMATH_GAUSSIAN1D_TCC

//# Includes
#include <casacore/scimath/Functionals/Gaussian1D.h>
#include <casacore/casa/BasicMath/Math.h>

namespace casacore { //# NAMESPACE CASACORE - BEGIN

//# Constructors

//# Operators
template<class T>
T Gaussian1D<T>::eval(typename Function1D<T>::FunctionArg x) const {
  T value = (x[0] - param_p[CENTER])/param_p[WIDTH]/fwhm2int;
  return param_p[HEIGHT] * exp(-(value*value));
}

//# Member functions

} //# NAMESPACE CASACORE - END


#endif
