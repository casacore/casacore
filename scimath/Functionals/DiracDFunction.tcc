//# DiracDFunction.cc: A one dimensional Dirac delta function
//# Copyright (C) 2002
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <casacore/scimath/Functionals/DiracDFunction.h>
#include <casacore/casa/BasicSL/Constants.h>
#include <casacore/casa/BasicMath/Math.h>

namespace casacore { //# NAMESPACE CASACORE - BEGIN

//# Constructors

//# Operators
template<class T> 
T DiracDFunction<T>::eval(typename Function<T>::FunctionArg x) const {
  T tmp(x[0] - param_p[CENTER]);
  if (tmp == T(0.0)) return param_p[HEIGHT];
  return T(0.0);
}

//# Member functions

} //# NAMESPACE CASACORE - END

