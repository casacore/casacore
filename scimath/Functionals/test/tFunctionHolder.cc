//# tFunctionHolder.cc: Test the one-dimensional scaled polynomial class
//# Copyright (C) 2002
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This program is free software; you can redistribute it and/or modify it
//# under the terms of the GNU General Public License as published by the Free
//# Software Foundation; either version 2 of the License, or (at your option)
//# any later version.
//#
//# This program is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
//# more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with this program; if not, write to the Free Software Foundation, Inc.,
//# 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <scimath/Functionals/FunctionHolder.h>

#include <scimath/Mathematics/AutoDiff.h>
#include <scimath/Mathematics/AutoDiffA.h>
#include <scimath/Mathematics/AutoDiffIO.h>
#include <scimath/Mathematics/AutoDiffMath.h>
#include <casa/BasicMath/Math.h>
#include <scimath/Functionals/UnaryFunction.h>
#include <scimath/Functionals/DiracDFunction.h>
#include <scimath/Functionals/GNoiseFunction.h>
#include <scimath/Functionals/KaiserBFunction.h>
#include <scimath/Functionals/SincFunction.h>
#include <scimath/Functionals/FunctionHolder.h>
#include <scimath/Functionals/Gaussian1D.h>
#include <casa/Utilities/Assert.h>

#include <casa/iostream.h>

#include <casa/namespace.h>
// Make near zero zero
Double Y(const Double in) {
  return (abs(in)<1e-15 ? Double(0.0) : in);
}

AutoDiff<Double> Y(AutoDiff<Double> in) {
  in.value() = Y(in.value());
  for (uInt i=0; i<in.nDerivatives(); ++i) {
    in.deriv(i) = Y(in.deriv(i));
  }
  return in;
}


int main() {
  cout << "---------------- test FunctionHolder ---------------" << endl;
  cout << "------------------------ Unary ---------------------" << endl;
  {
    UnaryFunction<Double> fn(5, 7, 3);
    for (Double x=3; x<11.2; x+=1.0) cout << "x=" << Y(x) << ": " <<
				       Y(fn(x)) << endl;
    UnaryFunction<AutoDiff<Double> > fnd;
    fnd[0] = AutoDiff<Double>(5, 3, 0);
    fnd[1] = AutoDiff<Double>(7, 3, 1);
    fnd[2] = AutoDiff<Double>(3, 3, 2);
    for (Double x=3; x<11.2; x+=1.0) cout << "x=" << Y(x) << ": " <<
				       Y(fnd(x)) << endl;
  }
  cout << "------------------------ Dirac delta ----------------" << endl;
  {
    DiracDFunction<Double> fn(5, 7);
    for (Double x=6; x<8.2; x+=1.0) cout << "x=" << Y(x) << ": " <<
				       Y(fn(x)) << endl;
    DiracDFunction<AutoDiff<Double> > fnd;
    fnd[0] = AutoDiff<Double>(5, 2, 0);
    fnd[1] = AutoDiff<Double>(7, 2, 1);
    for (Double x=6; x<8.2; x+=1.0) cout << "x=" << Y(x) << ": " <<
				      Y(fnd(x)) << endl;
  }
  cout << "------------------------ Normal noise ---------------" << endl;
  {
    GNoiseFunction<Double> fn(0, 2.0);
    for (uInt i=0; i<10; ++i) cout << fn() << endl;
    GNoiseFunction<AutoDiff<Double> > fnd;
    for (uInt i=0; i<10; ++i) cout << fnd() << endl;
  }
  cout << "------------------------ Kaiser-Bessel --------------" << endl;
  {
    KaiserBFunction<Double> fn;
    for (Double x=-1.2; x<1.21; x+=0.2) cout << "x=" << Y(x) << ": " <<
					  Y(fn(x)) << endl;
    for (Double x=6; x<8.2; x+=1.0) cout << "x=" << Y(x) << ": " <<
				      Y(fn(x)) << endl;
    KaiserBFunction<AutoDiff<Double> > fnd;
    fnd[0] = AutoDiff<Double>(1, 4, 0);
    fnd[1] = AutoDiff<Double>(0, 4, 1);
    fnd[2] = AutoDiff<Double>(1, 4, 2);
    fnd[3] = AutoDiff<Double>(2.5, 4, 2);
    for (Double x=-1.2; x<1.21; x+=0.2) cout << "x=" << Y(x) << ": " <<
					  Y(fnd(x)) << endl;
  }
  cout << "------------------------ sinc -----------------------" << endl;
  {
    SincFunction<Double> fn;
    for (Double x=-1.2; x<1.21; x+=0.2) cout << "x=" << Y(x) << ": " <<
					  Y(fn(x)) << endl;
    SincFunction<AutoDiff<Double> > fnd;
    fnd[0] = AutoDiff<Double>(1, 3, 0);
    fnd[1] = AutoDiff<Double>(0, 3, 1);
    fnd[2] = AutoDiff<Double>(1, 3, 2);
    for (Double x=-1.19999; x<1.21; x+=0.2) cout << "x=" << Y(x) << ": " <<
					      Y(fnd(x)) << endl;
  }
  cout << "------------------------ Gaussian1D -----------------" << endl;
  {
    Gaussian1D<Double> fn;
    FunctionHolder<Double> fh(fn);
    for (Double x=-1.2; x<1.21; x+=0.2) cout << "x=" << Y(x) << ": " <<
					  fh.asFunction()(x) << endl;
    Gaussian1D<AutoDiff<Double> > fnd;
    fnd[0] = AutoDiff<Double>(1, 3, 0);
    fnd[1] = AutoDiff<Double>(0, 3, 1);
    fnd[2] = AutoDiff<Double>(1, 3, 2);
    ///    FunctionHolder<AutoDiff<Double> > fhd(fnd);
    for (Double x=-1.2; x<1.21; x+=0.2) cout << "x=" << Y(x) << ": " <<
    ///					  fhd.asFunction()(x) << endl;
					  Y(fnd(x)) << endl;
  }

  cout << "-----------------------------------------------------" << endl;
  cout << "OK" << endl;
  return 0;
}
