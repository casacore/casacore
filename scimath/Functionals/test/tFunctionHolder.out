---------------- test FunctionHolder ---------------
------------------------ Unary ---------------------
x=3: 0
x=4: 2.5
x=5: 5
x=6: 5
x=7: 5
x=8: 5
x=9: 5
x=10: 2.5
x=11: 0
x=3: (0, [])
x=4: (2.5, [0.5, 0, 0])
x=5: (5, [1, 0, 0])
x=6: (5, [1, 0, 0])
x=7: (5, [1, 0, 0])
x=8: (5, [1, 0, 0])
x=9: (5, [1, 0, 0])
x=10: (2.5, [0.5, 0, 0])
x=11: (0, [])
------------------------ Dirac delta ----------------
x=6: 0
x=7: 5
x=8: 0
x=6: (0, [])
x=7: (5, [1, 0])
x=8: (0, [])
------------------------ Normal noise ---------------
-0.569721
-2.0255
2.0087
-0.445608
0.302747
-0.687715
-0.892452
0.302263
-1.51307
-0.610897
(-0.402854, [])
(-1.43224, [])
(1.42036, [])
(-0.315093, [])
(0.214074, [])
(-0.486288, [])
(-0.631059, [])
(0.213732, [])
(-1.0699, [])
(-0.43197, [])
------------------------ Kaiser-Bessel --------------
x=-1.2: 0.0879273
x=-1: 0.0227113
x=-0.8: 0.0879273
x=-0.6: 0.286772
x=-0.4: 0.588771
x=-0.2: 0.87857
x=0: 1
x=0.2: 0.87857
x=0.4: 0.588771
x=0.6: 0.286772
x=0.8: 0.0879273
x=1: 0.0227113
x=1.2: 0.0879273
x=6: 1
x=7: 0.0227113
x=8: 1
x=-1.2: (0.0879273, [0.0879273, 0.659253, -0.897684, 0])
x=-1: (0.0227113, [0.0227113, 0, -0.0597192, 0])
x=-0.8: (0.0879273, [0.0879273, -0.659253, 0.420821, 0])
x=-0.6: (0.286772, [0.286772, -1.30976, 0.609155, 0])
x=-0.4: (0.588771, [0.588771, -1.61021, 0.485836, 0])
x=-0.2: (0.87857, [0.87857, -1.14539, 0.170811, 0])
x=0: (1, [1, 0, 0, 0])
x=0.2: (0.87857, [0.87857, 1.14539, 0.170811, 0])
x=0.4: (0.588771, [0.588771, 1.61021, 0.485836, 0])
x=0.6: (0.286772, [0.286772, 1.30976, 0.609155, 0])
x=0.8: (0.0879273, [0.0879273, 0.659253, 0.420821, 0])
x=1: (0.0227113, [0.0227113, 0, -0.0597192, 0])
x=1.2: (0.0879273, [0.0879273, -0.659253, -0.897684, 0])
------------------------ sinc -----------------------
x=-1.2: -0.155915
x=-1: 0
x=-0.8: 0.233872
x=-0.6: 0.504551
x=-0.4: 0.756827
x=-0.2: 0.935489
x=0: 1
x=0.2: 0.935489
x=0.4: 0.756827
x=0.6: 0.504551
x=0.8: 0.233872
x=1: 0
x=1.2: -0.155915
x=-1.19999: (-0.155909, [0, -0.544276, 0.653126])
x=-0.99999: (1.00001e-05, [0, -1.00002, 1.00001])
x=-0.79999: (0.233885, [0, -1.30362, 1.04288])
x=-0.59999: (0.504565, [0, -1.35594, 0.813552])
x=-0.39999: (0.756838, [0, -1.11951, 0.447791])
x=-0.19999: (0.935496, [0, -0.632332, 0.12646])
x=1e-05: (1, [0, 3.28987e-05, 3.28987e-10])
x=0.20001: (0.935483, [0, 0.632391, 0.126484])
x=0.40001: (0.756816, [0, 1.11954, 0.447828])
x=0.60001: (0.504538, [0, 1.35595, 0.813584])
x=0.80001: (0.233859, [0, 1.3036, 1.04289])
x=1.00001: (-9.9999e-06, [0, 0.99998, 0.99999])
x=1.20001: (-0.15592, [0, 0.544227, 0.653078])
------------------------ Gaussian1D -----------------
x=-1.2: 0.018453
x=-1: 0.0625
x=-0.8: 0.169576
x=-0.6: 0.368567
x=-0.4: 0.641713
x=-0.2: 0.895025
x=0: 1
x=0.2: 0.895025
x=0.4: 0.641713
x=0.6: 0.368567
x=0.8: 0.169576
x=1: 0.0625
x=1.2: 0.018453
x=-1.2: (0.018453, [0.018453, -0.12279, 0.147348])
x=-1: (0.0625, [0.0625, -0.346574, 0.346574])
x=-0.8: (0.169576, [0.169576, -0.752261, 0.601809])
x=-0.6: (0.368567, [0.368567, -1.22626, 0.735758])
x=-0.4: (0.641713, [0.641713, -1.42336, 0.569346])
x=-0.2: (0.895025, [0.895025, -0.992615, 0.198523])
x=0: (1, [1, 0, 0])
x=0.2: (0.895025, [0.895025, 0.992615, 0.198523])
x=0.4: (0.641713, [0.641713, 1.42336, 0.569346])
x=0.6: (0.368567, [0.368567, 1.22626, 0.735758])
x=0.8: (0.169576, [0.169576, 0.752261, 0.601809])
x=1: (0.0625, [0.0625, 0.346574, 0.346574])
x=1.2: (0.018453, [0.018453, 0.12279, 0.147348])
-----------------------------------------------------
OK
